galaxyproject/galaxy
# Pulsar infrastructure improvements for TES execution and MQ-less Kubernetes.

Updating the Pulsar job runner/client for https://github.com/galaxyproject/pulsar/pull/302 and as described in the new docs at https://pulsar.readthedocs.io/en/latest/containers.html.

The open PR for TES integration (https://github.com/galaxyproject/galaxy/pull/14462) is a fantastic attempt @vipulchhabra99 - the approach to TES in this PR is just a bit stronger though. This approach heavily borrows on that work however and I'm very grateful for the impressive GSOC project outcome.

The proposed TES runner in #14462 does what I have been calling co-execution - that is process of running a piece of Pulsar/Galaxy/etc... code beside the target container to do staging. The approach I have integrated with this work reuses the Kubernetes co-execution container target instead of a new though which has more automated test coverage (even if I haven't succeeded with integrating it into the CI yet) and results in one less thing we need to maintain. The container referenced here has already been published to Galaxy's Docker Hub. The container also does a much more expansive job staging files up and down and is capable of performing extended metadata collection. I think as a result of piggy-backing on the Pulsar approach the resulting runner will correctly run more kinds of tools, tool features (index data, configuration files, Conda, etc..), and under more Galaxy and related configuration modalities.

Rather than using ``requests`` directly - this approach is based on pydantic-tes (https://github.com/jmchilton/pydantic-tes) - which uses a Pydantic validated client and richly typed models. A benefit of using this library to interact with TES - is that the library has its own CI and runs tests against a real funnel server so we have strong test coverage that runs in isolation for all of those interactions. Changes to Funnel, changes to the schema, the addition of a new TES target, adding authorization mechanisms, etc... can all be done and tested isolation without concerns about Pulsar or Galaxy.

Mirroring the Kubernetes approach, this approach uses Pulsar as a client for staging Galaxy jobs up and down but does not require a Pulsar server at all - TES is responsible for persisting the execution state, submitting jobs, etc.... The new test cases included describe Galaxy job configurations for both containerized tool jobs and conda-based jobs using TES.

## Different Deployment Scenarios

Using a tool container and a message queue for communicating final job status:

![](https://pulsar.readthedocs.io/en/latest/_images/pulsar_tes_coexecution_mq_deployment.plantuml.svg)

Using the Pulsar staging container + conda for dependency management along with a message queue for communicating final job status:

![](https://pulsar.readthedocs.io/en/latest/_images/pulsar_tes_mq_deployment.plantuml.svg)

Using a tool container but no message queue for communicating final job status - instead relying on extended metadata.

![](https://pulsar.readthedocs.io/en/latest/_images/pulsar_tes_coexecution_deployment.plantuml.svg)

Using the Pulsar staging container + conda for dependency management but no message queue for communicating final job status - instead relying on extended metadata.

![](https://pulsar.readthedocs.io/en/latest/_images/pulsar_tes_deployment.plantuml.svg)



## How to test the changes?
(Select all options that apply)
  1. Configure your machine to run Docker.
  2.  Run a funnel server on port 8000 - run pydantic-tes test suite against it to ensure responses validate and containers launch.
  3.  Run Rabbit MQ server - expose it on 0.0.0.0 by setting NODE_IP_ADDRESS= to empty string in the Rabbit MQ configuration.
  4. ``GALAXY_TEST_PORT=9234 GALAXY_TEST_AMQP_URL="amqp://guest:guest@localhost:5672" GALAXY_TEST_INFRASTRUCTURE_HOST="_PLATFORM_AUTO_"   FUNNEL_SERVER_TARGET=http://localhost:8000 pytest -s test/integration/test_coexecution.py::TestTesCoexecutionContainerIntegration``

## License
